### PR TITLE
API Docs Generator

### PR DESCRIPTION
**Pull request by jamiebrynes7**
Okay this is _probably_ ready for a preliminary review.

I've tried to break it down into logical commits, but some parts reference future commits but it _should_ be fine to just look at the API call name and infer what it does. Otherwise it needs a better name!

### CODE REVIEW

File: tools/Docgen/Builder/DoxygenParser.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+using System.Xml.Serialization;
+using Docgen.Xml;
+
+namespace Docgen.Builder
+{
+    public static class DoxygenParser
+    {
+        private static List<string> BlackList = new List<string>
+        {
+            "index.xml"
```
Comment (jessicafalk) [inline tools/Docgen/Builder/DoxygenParser.cs:13]: maybe add a short comment on why this needs to be blacklisted

File: tools/Docgen/Utils/Formatting.cs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        /// </summary>
+        /// <remarks>
+        ///     Note that the relative path has the target namespace stripped off the front of it.
+        /// </remarks>
+        /// <param name="compoundName">The Doxygen compound name.</param>
+        /// <returns>The relative path from the root of the API docs output folder.</returns>
+        public static string CompoundNameToRelativePath(string compoundName)
+        {
+            var targetNamespace = Options.Instance.TargetNamespace.Replace(".", "::") + "::";
+            var markdownSplitPath = compoundName.Replace(targetNamespace, "").Split("::")
```
Comment (jessicafalk) [inline tools/Docgen/Utils/Formatting.cs:36]: use string.Empty

File: tools/Docgen/Utils/Formatting.cs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        }
+        
+        /// <summary>
+        ///     Selects the type name from a full qualified type name.
+        /// </summary>
+        /// <param name="fullyQualified">The fully qualified type name. Expected in the format of Foo.Bar.MyClass</param>
+        /// <returns>The type name.</returns>
+        public static string GetTypeFromFullyQualified(string fullyQualified)
+        {
+            return fullyQualified.Split(".").Reverse().First();
```
Comment (jessicafalk) [inline tools/Docgen/Utils/Formatting.cs:73]: why not just use Last()?

Comment (jamiebrynes7) [reply tools/Docgen/Utils/Formatting.cs:73]: Ah - yes

File: tools/Docgen/Builder/Builder.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        public static void Build(DoxygenDatabase database)
+        {
+            CleanOutputDir(Options.Instance.OutputDirectory);
+            BuildDocs(database, Options.Instance.OutputDirectory);
+            BuildTableOfContents(database, Options.Instance.OutputDirectory);
+        }
+
+        private static void BuildTableOfContents(DoxygenDatabase database, string outputDirectory)
+        {
+            // Build the table of contents at {{outputDirectory}}/.toc/toc.md
```
Comment (jessicafalk) [inline tools/Docgen/Builder/Builder.cs:17]: that is the toc just for the API docs right? how would this work together with the rest of the documentation?

File: tools/Docgen/Builder/Builder.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            var filepath = Path.Combine(outputDirectory, ".toc", "toc.md");
+            Directory.CreateDirectory(Path.GetDirectoryName(filepath));
+            File.WriteAllText(filepath, new TableOfContentsGenerator(database).Generate());
+        }
+
+        private static void BuildDocs(DoxygenDatabase database, string outputDirectory)
+        {            
+            foreach (var details in database.TypeDetails)
```
Comment (jessicafalk) [inline tools/Docgen/Builder/Builder.cs:25]: those loops could potentially be factored out into a method.

File: tools/Docgen/Builder/Builder.cs (hunk 4 above)
Comment (jamiebrynes7) [reply tools/Docgen/Builder/Builder.cs:17]: Yes and you can use the include macro in the main ToC.

File: tools/Docgen/Builder/Builder.cs (hunk 5 above)
Comment (jamiebrynes7) [reply tools/Docgen/Builder/Builder.cs:25]: Yeah agreed.

**Comment by jamiebrynes7** [comment]
Moved to another repo

**Pull request closed without merging**